File: cli/src/agent/tool-host/pi-extension.ts
/**
 * Locate and verify the bundled Cognia Pi extension (ADR-0119).
 *
 * The extension is what enforces Pi's native-tool permission matrix, so two
 * things have to be true before a session may start: the file must be found,
 * and it must be the file Cognia shipped. A tampered or substituted extension
 * would silently hold the permission gate open, and the handshake alone cannot
 * detect that — a modified extension can still announce itself.
 *
 * Resolution mirrors `resolveToolBridgeScript` so both packaged artifacts are
 * discovered the same way:
 *   1. `$COGNIA_PI_EXTENSION_PATH` (explicit override)
 *   2. a `sidecar/` dir next to the executable (packaged dist layout)
 *   3. a `sidecar/` dir walked up from this module (in-repo / dev)
 */

import { createHash } from "node:crypto"
import fs from "node:fs"
import path from "node:path"
import { fileURLToPath } from "node:url"

const EXTENSION_RELATIVE = path.join("sidecar", "pi-extension", "cognia-pi-extension.ts")
const INTEGRITY_RELATIVE = path.join("sidecar", "pi-extension", "integrity.json")

export interface ResolvePiExtensionOptions {
  env?: NodeJS.ProcessEnv
  exists?: (candidate: string) => boolean
  execPath?: string
}

/**
 * Is `$COGNIA_PI_EXTENSION_PATH` allowed to redirect the resolver?
 *
 * Only outside production. The override exists so a contributor can iterate on
 * the extension without re-pinning after every edit; in a shipped build it is
 * an env var that swaps out the component holding the permission gate, which
 * is precisely the substitution the digest pin exists to prevent.
 */
export function piExtensionOverrideAllowed(env: NodeJS.ProcessEnv = process.env): boolean {
  return env.NODE_ENV !== "production"
}

/** Absolute path of the bundled extension, or undefined when it is absent. */
export function resolvePiExtensionScript(opts: ResolvePiExtensionOptions = {}): string | undefined {
  const env = opts.env ?? process.env
  const exists = opts.exists ?? fs.existsSync
  const execPath = opts.execPath ?? process.execPath

  const override = env.COGNIA_PI_EXTENSION_PATH?.trim()
  // An explicit override is returned even when it does not exist, so a typo
  // surfaces as "the file you pointed at is missing" rather than silently
  // falling back to a different extension than the operator intended.
  if (override && piExtensionOverrideAllowed(env)) return override

  const adjacent = path.join(path.dirname(execPath), EXTENSION_RELATIVE)
  if (exists(adjacent)) return adjacent

  let dir: string
  try {
    dir = path.dirname(fileURLToPath(import.meta.url))
  } catch {
    dir = typeof __dirname === "string" ? __dirname : process.cwd()
  }
  for (let i = 0; i < 10; i++) {
    const candidate = path.join(dir, EXTENSION_RELATIVE)
    if (exists(candidate)) return candidate
    const parent = path.dirname(dir)
    if (parent === dir) break
    dir = parent
  }
  return undefined
}

/** SHA-256 of a file's bytes, lowercase hex. */
export function digestFile(file: string, read: (p: string) => Buffer = fs.readFileSync): string {
  return createHash("sha256").update(read(file)).digest("hex")
}

export interface PiExtensionIntegrity {
  /** Expected SHA-256 of `cognia-pi-extension.ts`. */
  sha256: string
}

export type PiExtensionVerdict =
  | { status: "ok"; path: string; sha256: string }
  | { status: "missing" }
  | { status: "unreadable"; path: string; detail: string }
  /** Found, but not the file Cognia shipped. */
  | { status: "tampered"; path: string; expected: string; actual: string }
  /** Found, but no pinned digest to compare against. */
  | { status: "unpinned"; path: string; sha256: string }

export interface VerifyPiExtensionOptions extends ResolvePiExtensionOptions {
  readFile?: (p: string) => Buffer
  readIntegrity?: (p: string) => string
}

/**
 * Resolve the extension and check it against its pinned digest.
 *
 * `unpinned` is reported rather than treated as success: shipping without the
 * manifest is a packaging mistake, and silently accepting any file there would
 * make the pin decorative.
 */
export function verifyPiExtension(opts: VerifyPiExtensionOptions = {}): PiExtensionVerdict {
  const resolved = resolvePiExtensionScript(opts)
  if (!resolved) return { status: "missing" }

  const readFile = opts.readFile ?? fs.readFileSync
  const readIntegrity = opts.readIntegrity ?? ((p: string) => fs.readFileSync(p, "utf8"))

  let actual: string
  try {
    actual = digestFile(resolved, readFile)
  } catch (error) {
    return { status: "unreadable", path: resolved, detail: String(error) }
  }

  const manifestPath = path.join(
    resolved.slice(0, resolved.length - EXTENSION_RELATIVE.length),
    INTEGRITY_RELATIVE
  )

  let expected: string | undefined
  try {
    const parsed = JSON.parse(readIntegrity(manifestPath)) as Partial<PiExtensionIntegrity>
    expected = typeof parsed.sha256 === "string" ? parsed.sha256.toLowerCase() : undefined
  } catch {
    expected = undefined
  }

  if (!expected) return { status: "unpinned", path: resolved, sha256: actual }
  if (expected !== actual) return { status: "tampered", path: resolved, expected, actual }
  return { status: "ok", path: resolved, sha256: actual }
}

/**
 * The path to hand the Pi adapter, or undefined when the extension must not be
 * loaded.
 *
 * **Only `ok` yields a path.** `unpinned` used to be accepted too, which made
 * the pin decorative: an extension shipped without its manifest — or with the
 * manifest stripped — loaded anyway and took charge of the permission gate.
 * "We could not verify this" and "this is verified" must not lead to the same
 * place when the component being verified is the thing enforcing permissions.
 *
 * A refusal is not a downgrade to "run without the extension": the caller
 * treats a missing path as a refused session (`PI_EXTENSION_REQUIRED`),
 * because without interception Pi's native tools run with the full rights of
 * the process.
 */
export function piExtensionPathForSpawn(verdict: PiExtensionVerdict): string | undefined {
  return verdict.status === "ok" ? verdict.path : undefined
}

/**
 * Why a session may not start, phrased for the user, or `undefined` when the
 * extension verified clean.
 */
export function piExtensionRefusalReason(verdict: PiExtensionVerdict): string | undefined {
  switch (verdict.status) {
    case "ok":
      return undefined
    case "missing":
      return "The bundled Cognia Pi extension was not found. Reinstall Cognia — without it Pi's native tools would run unintercepted."
    case "unreadable":
      return `The bundled Cognia Pi extension at ${verdict.path} could not be read: ${verdict.detail}`
    case "tampered":
      return `The bundled Cognia Pi extension at ${verdict.path} does not match its pinned digest (expected ${verdict.expected}, found ${verdict.actual}). Reinstall Cognia.`
    case "unpinned":
      return `The bundled Cognia Pi extension at ${verdict.path} has no integrity manifest, so it cannot be verified. Reinstall Cognia.`
  }
}

File: src-tauri/src/pi_extension.rs
//! Desktop resolution + integrity check for the bundled Cognia Pi extension
//! (ADR-0119).
//!
//! The extension is what intercepts `pi.on("tool_call")` and applies Cognia's
//! permission matrix to Pi's native `read`/`edit`/`write`/`bash` tools. Pi ships
//! no permission prompts of its own, so without the extension those tools run
//! with the full rights of the process. That makes "is this the extension we
//! shipped?" a precondition for starting a session, not a diagnostic.
//!
//! The CLI answers the same question in TypeScript
//! (`cli/src/agent/tool-host/pi-extension.ts`). The desktop cannot call into
//! that code — the renderer is a static export with no filesystem — so the
//! adapter asks its host instead, and each host answers with the same verdict
//! shape. Keeping one *contract* with two implementations is what lets the
//! adapter stay host-agnostic; keeping one *implementation* is not possible
//! here.
//!
//! Resolution deliberately reuses [`crate::claude::sidecar::sidecar_dir`], so
//! the packaged (`resource_dir/sidecar`, `resource_dir/_up_/sidecar`) and
//! checkout layouts are discovered exactly where every other sidecar asset is.

use std::path::{Path, PathBuf};

use serde::Serialize;
use sha2::{Digest, Sha256};
use tauri::AppHandle;

/// Verdict for one resolution attempt. Serialized to the renderer with
/// `camelCase` keys so it deserializes as the same union the CLI produces.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(tag = "status", rename_all = "camelCase")]
pub enum PiExtensionVerdict {
    /// Found, and the digest matches the pinned manifest.
    #[serde(rename_all = "camelCase")]
    Ok { path: String, sha256: String },
    /// No extension on disk.
    Missing,
    /// Present but unreadable.
    #[serde(rename_all = "camelCase")]
    Unreadable { path: String, detail: String },
    /// Present, but not the bytes Cognia shipped.
    #[serde(rename_all = "camelCase")]
    Tampered {
        path: String,
        expected: String,
        actual: String,
    },
    /// Present, but there is no manifest to compare against.
    ///
    /// Reported rather than accepted: an extension whose digest cannot be
    /// checked is exactly the substitution the pin exists to catch, and
    /// "unverifiable" must not lead to the same place as "verified".
    #[serde(rename_all = "camelCase")]
    Unpinned { path: String, sha256: String },
}

fn digest_bytes(bytes: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(bytes);
    hex::encode(hasher.finalize())
}

/// Verify the extension inside an already-resolved `sidecar/` directory.
///
/// Split from the Tauri command so the whole matrix is testable without an
/// `AppHandle`.
pub fn verify_in_sidecar_dir(sidecar_dir: &Path) -> PiExtensionVerdict {
    let extension = sidecar_dir
        .join("pi-extension")
        .join("cognia-pi-extension.ts");
    if !extension.is_file() {
        return PiExtensionVerdict::Missing;
    }
    let path = extension.display().to_string();

    let actual = match std::fs::read(&extension) {
        Ok(bytes) => digest_bytes(&bytes),
        Err(error) => {
            return PiExtensionVerdict::Unreadable {
                path,
                detail: error.to_string(),
            }
        }
    };

    let manifest = sidecar_dir.join("pi-extension").join("integrity.json");
    let expected = std::fs::read_to_string(&manifest)
        .ok()
        .and_then(|raw| serde_json::from_str::<serde_json::Value>(&raw).ok())
        .and_then(|value| {
            value
                .get("sha256")
                .and_then(|v| v.as_str())
                .map(|v| v.to_ascii_lowercase())
        });

    match expected {
        None => PiExtensionVerdict::Unpinned {
            path,
            sha256: actual,
        },
        Some(expected) if expected != actual => PiExtensionVerdict::Tampered {
            path,
            expected,
            actual,
        },
        Some(_) => PiExtensionVerdict::Ok {
            path,
            sha256: actual,
        },
    }
}

/// Resolve and verify the bundled Pi extension on this desktop host.
///
/// Returns a verdict rather than an error for every "we looked and it is not
/// usable" case: the adapter turns that into a refused session with a message
/// naming the cause, which `extension_handshake_failed` could not do (it could
/// not tell "never shipped" from "tampered" from "timed out").
#[tauri::command]
pub fn resolve_pi_extension(app: AppHandle) -> Result<PiExtensionVerdict, String> {
    let dir: PathBuf = crate::claude::sidecar::sidecar_dir(&app)?;
    Ok(verify_in_sidecar_dir(&dir))
}

/// The sidecar directory a headless Host ships its assets in.
///
/// `COGNIA_SIDECAR_SCRIPT` names the host entry, so its parent is the
/// directory, and the checkout fallback matches `resolve_sidecar_script_path`
/// in `bin/cognia-server.rs` so a hand-run `serve` finds it too. Kept beside
/// the verifier rather than reaching into the binary, because a library arm
/// cannot call into a `bin` target.
pub fn headless_sidecar_dir() -> Option<PathBuf> {
    if let Ok(raw) = std::env::var("COGNIA_SIDECAR_SCRIPT") {
        let trimmed = raw.trim();
        if !trimmed.is_empty() {
            if let Some(parent) = Path::new(trimmed).parent() {
                return Some(parent.to_path_buf());
            }
        }
    }
    let checkout = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()?
        .join("sidecar");
    checkout.is_dir().then_some(checkout)
}

/// The verdict for whichever host is answering.
///
/// The Pi adapter runs in a renderer and asks its host this question. A browser
/// paired to a Host is that renderer, and it has no filesystem at all, so the
/// question has to reach the machine the Pi process will actually run on. That
/// is why this is host-neutral rather than a second desktop-only path.
pub fn verify_for_host(sidecar_dir: Option<PathBuf>) -> PiExtensionVerdict {
    match sidecar_dir {
        Some(dir) => verify_in_sidecar_dir(&dir),
        None => PiExtensionVerdict::Missing,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn write(dir: &Path, name: &str, contents: &str) {
        let target = dir.join("pi-extension");
        std::fs::create_dir_all(&target).expect("mkdir");
        std::fs::write(target.join(name), contents).expect("write");
    }

    #[test]
    fn missing_extension_is_reported_not_guessed() {
        let tmp = tempfile::tempdir().expect("tempdir");
        assert_eq!(
            verify_in_sidecar_dir(tmp.path()),
            PiExtensionVerdict::Missing
        );
    }

    #[test]
    fn a_matching_digest_verifies() {
        let tmp = tempfile::tempdir().expect("tempdir");
        write(tmp.path(), "cognia-pi-extension.ts", "export const x = 1\n");
        let sha = digest_bytes(b"export const x = 1\n");
        write(
            tmp.path(),
            "integrity.json",
            &format!("{{\"sha256\":\"{sha}\"}}"),
        );

        match verify_in_sidecar_dir(tmp.path()) {
            PiExtensionVerdict::Ok { sha256, .. } => assert_eq!(sha256, sha),
            other => panic!("expected ok, got {other:?}"),
        }
    }

    /// The case the pin exists for: a substituted extension would otherwise
    /// take charge of the permission gate.
    #[test]
    fn an_edited_extension_is_tampered() {
        let tmp = tempfile::tempdir().expect("tempdir");
        write(tmp.path(), "cognia-pi-extension.ts", "malicious()\n");
        write(
            tmp.path(),
            "integrity.json",
            &format!("{{\"sha256\":\"{}\"}}", digest_bytes(b"original\n")),
        );

        match verify_in_sidecar_dir(tmp.path()) {
            PiExtensionVerdict::Tampered {
                expected, actual, ..
            } => assert_ne!(expected, actual),
            other => panic!("expected tampered, got {other:?}"),
        }
    }

    /// Stripping the manifest must not be a way to bypass the pin.
    #[test]
    fn a_missing_or_malformed_manifest_is_unpinned_never_ok() {
        let tmp = tempfile::tempdir().expect("tempdir");
        write(tmp.path(), "cognia-pi-extension.ts", "export const x = 1\n");
        assert!(matches!(
            verify_in_sidecar_dir(tmp.path()),
            PiExtensionVerdict::Unpinned { .. }
        ));

        write(tmp.path(), "integrity.json", "not json at all");
        assert!(matches!(
            verify_in_sidecar_dir(tmp.path()),
            PiExtensionVerdict::Unpinned { .. }
        ));

        write(tmp.path(), "integrity.json", "{\"sha256\": 42}");
        assert!(matches!(
            verify_in_sidecar_dir(tmp.path()),
            PiExtensionVerdict::Unpinned { .. }
        ));
    }

    #[test]
    fn a_pinned_digest_compares_case_insensitively() {
        let tmp = tempfile::tempdir().expect("tempdir");
        write(tmp.path(), "cognia-pi-extension.ts", "export const x = 1\n");
        let sha = digest_bytes(b"export const x = 1\n").to_uppercase();
        write(
            tmp.path(),
            "integrity.json",
            &format!("{{\"sha256\":\"{sha}\"}}"),
        );
        assert!(matches!(
            verify_in_sidecar_dir(tmp.path()),
            PiExtensionVerdict::Ok { .. }
        ));
    }

    /// The renderer deserializes this as the same tagged union the CLI emits.
    #[test]
    fn verdicts_serialize_with_the_shared_tagged_shape() {
        let value = serde_json::to_value(PiExtensionVerdict::Ok {
            path: "/x/cognia-pi-extension.ts".into(),
            sha256: "abc".into(),
        })
        .expect("serialize");
        assert_eq!(value["status"], "ok");
        assert_eq!(value["path"], "/x/cognia-pi-extension.ts");
        assert_eq!(value["sha256"], "abc");

        let missing = serde_json::to_value(PiExtensionVerdict::Missing).expect("serialize");
        assert_eq!(missing["status"], "missing");
    }

    /// The reason this became host-neutral: a browser paired to a Host is the
    /// renderer that asks, and it has no filesystem at all. Answering
    /// `Missing` for "we could not find a sidecar directory" keeps the refusal
    /// honest instead of reporting a verified extension nobody looked at.
    #[test]
    fn a_host_with_no_sidecar_directory_reports_missing() {
        assert_eq!(verify_for_host(None), PiExtensionVerdict::Missing);
    }

    #[test]
    fn a_host_with_a_sidecar_directory_gets_that_directory_s_verdict() {
        let tmp = tempfile::tempdir().expect("tempdir");
        write(tmp.path(), "cognia-pi-extension.ts", "export default {}\n");
        write(
            tmp.path(),
            "integrity.json",
            &format!(
                "{{\"sha256\":\"{}\"}}",
                digest_bytes(b"export default {}\n")
            ),
        );

        assert!(matches!(
            verify_for_host(Some(tmp.path().to_path_buf())),
            PiExtensionVerdict::Ok { .. }
        ));
    }

    #[test]
    fn the_headless_sidecar_directory_is_the_host_script_s_parent() {
        let tmp = tempfile::tempdir().expect("tempdir");
        let script = tmp.path().join("agent-host.mjs");
        std::fs::write(&script, "").expect("write");
        // Serialised with the other env-reading test by construction: this is
        // the only one that sets it, and it clears it before returning.
        std::env::set_var("COGNIA_SIDECAR_SCRIPT", &script);
        let resolved = headless_sidecar_dir();
        std::env::remove_var("COGNIA_SIDECAR_SCRIPT");

        assert_eq!(resolved.as_deref(), Some(tmp.path()));
    }
}
